### DISCUSSION TITLE
Rules Advice - Tasmota on Shelly 1 as offgrid pump controller

### DISCUSSION THREAD

**Discussion by blackshoals** [arendst/Tasmota#15901] [Q&A] [answered]
I have an offgrid cabin that I get to once or twice a year. Power is 12 volt solar from batteries and water is rainwater piped to a collector tank  then pumped to a main tank. My current system uses a float switch in the collector tank and a voltage disconnect to keep the batteries from being run down. In rainy season that means the pump is running every few hours as long as the batteries are up. It will also try to run if the weather is cold. There is no internet connection other than phone hotspot when I'm there. I want the pump to run  about every 24 hours for about 20 minutes if the temperature is warm enough. If the main tank is full overpumping is OK as it keeps the water mixed a bit. I know without a RTC and no NTP connection Tasmota timing will drift but the interval is not critical. It has to perform when if the batteries cutoff and on periodically.

I have a Shelly 1 that can handle 15A and have flashed Tasmota and added a DS18B20 temperature probe. I've created 3 rules:

#On boot (ie batteries have recharged) set the time to  June 29 2021, relay off, and a 1 hour delay to let batteries and the Shelly settle before the pump timer cycle (Rule2) starts. Turn off the wifi after 1 hour to save power/scanning. It will enable itself on a reboot giving a window to make changes from my phone hotspot.

**Rule1** ON Power1#Boot DO Backlog Time 1656543600; PowerOnState 0; Rule2 0; RuleTimer1 3600 ENDON ON Rules#Timer=1 DO Backlog Rule2 1; Wifi 0 ENDON

#Approximately every 24 hours turn the pump on for 20 minutes

**Rule2** ON Time#Hour|24 DO Backlog Power1 On; RuleTimer2 1200 ENDON ON Rules#Timer=2 DO Power1 off ENDON

#If the temp is below 9C turn off pump

**Rule3** ON DS18B20#Temperature < 9 DO Power1 Off ENDON

Doing some testing the Shelly seems to perform OK. For the manual override I won't use the Shelly manual switch terminals but will bypass it altogether and the float switch will remain in series with the pump as a failsafe against a stuck relay or software problem. With the remote location I'd be grateful if anyone sees a problem with my usecase. Thanks

**Comment by barbudor** [comment, marked as answer]
I see aproblem with rule 2 as the trigger `Time#Hour` doesn't exist. Check existing triggers at https://tasmota.github.io/docs/Rules/#rule-trigger
However if your idea is to trigger once a day, you can use `Time#Minute=xxx` where `xxx` is the number of minutes since 00:00
Let say you want to trigger the rule everyday at 8:30 : xxx = 8*60+30 = 510

You didn't say how you enabled each rules.
I was about to suggest to use "once" mode on Rule 3 (using `Rule3 5`) however this would be bad as if the low temp threshold is crossed before the pump switch time, it would not stop it. However it will flood the console with messages as it will try to switch power off continuously once or more per second

A nicer way would be to use somehting like:
```
rule3 ON DS18B20#Temperature < 9 DO backlog var1 0; Power1 Off ENDON on DS18B20#Temperature > 9  DO var1 1 ENDON
Rule3 1
Rule3 5

Rule2 ON Time#Minute=510 DO Backlog Power1 var1; RuleTimer2 1200 ENDON ON Rules#Timer=2 DO Power1 off ENDON
Rule2 1

Rule1 ON Power1#Boot DO Backlog Time 1656543600; PowerOnState 0; Rule2 0; RuleTimer1 3600; var1 1 ENDON ON Rules#Timer=1 DO Backlog Rule2 1; Wifi 0 ENDON
```
Which will prevent Rule2 to power on the pump while the temperature is below the threshold while avoiding the message flood


I understand your need for low power which means no wifi, no time server... but this is not what Tasmota was designed for so don't be surprised if it behave erratically.

**Reply by sfromis** [reply]
Also beware that, in an environment where power may fluctuate, the MCU just might get in a state where it will be frozen, not necessarily remedied by power getting "good" again. Of course, this depends on how your power supply works, but with a commercial device, it can be hard to tell if the hardware design will do what it "should", completely cut off MCU power when a brownout threatens. Of course, you should more generally consider how bad it would be for the controller suddenly going dead for whatever reason.

**Reply by blackshoals** [reply]
Thank you sfromis, it's an important point. The upstream voltage is supplied by a commercial regulator that cuts off to 0 when the batteries can no longer supply 12 volts, then reconnects at 12 volts when available so a brownout shouldn't be an issue. I haven't put a scope on it to test it but the company has a good track record.

**Reply by blackshoals** [reply]
Thank you barbudor. Oddly enough Time#Hour did seem to test OK although you are correct about the documentation. I enabled Rules as 1, and Rule 3, the temperature protection, overrides the call from Rule 2 for the pump to come on, but it does try to cycle then shut down. Your solution looks more elegant.The appeal of the Shelly is that it is a ready made package of an ESP8266 and high current relay that is 12 volt compatible, as opposed to soldering an Arduino, voltage regulator and relay together. An Arduino sketch would likely fit better here than Tasmota, but after a quick scan I couldn't find an example of anyone working with one.

**Reply by blackshoals** [reply]
Small edit to the second inequality in Rule 3 - should be ">=" vs "<"

rule3 ON DS18B20#Temperature < 9 DO backlog var1 0; Power1 Off ENDON on DS18B20#Temperature >= 9  DO var1 1 ENDON

**Reply by barbudor** [reply]
🤦‍♂️
I've edited the original one for anyone to get right for the 1st time
I prefer ">" to leave a small hysteresis

**Reply by blackshoals** [reply]
Thanks, the hysteresis buffer makes sense. Another minor edit is that the reference to var1 in Rule 2 needs to be escaped by % otherwise it seems to pass as text and was not turning on. Not sure if that's related to my text editor. Tests OK so far.

Rule2 ON Time#Minute=510 DO Backlog Power1 %var1%; RuleTimer2 1200 ENDON ON Rules#Timer=2 DO Power1 off ENDON

**Comment by shwanaa** [comment]
hello how i can mame rulle on tasmota if  the first switch on this second off
